org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/DOMCompletionEngine.java

```diff
@@ -0,0 +1,188 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.jdt.internal.core;
```

Comment:
i suggest to bundle all DOM Implemenations in a separate package or even bundle

Reply:
Better have new API (getter), if necessary, then to have the different implementations spread around the whole code base. The goal should be that if the experiment should go wrong, it is enough to delete the alternative implementation folder. JDT had the same issue with the "new indexer", and still there are references in the code, because they where not easy to find :-(

Reply:
What you're asking for is a major refactoring of existing JDT code: a lot of code to move, a lot of interfaces/methods would then become public APIs. This is beyond the current resources of the project and beyond what is necessary for innovating on those aspects.